src/common/position_path.rs

```diff
@@ -0,0 +1,550 @@
+/// returns a tuple containing the next position in the path and the corresponding side node.
+/// Because the tree may be imbalanced, as described by the path's `leaves_count` parameter, the
+/// side node may not necessarily be the direct sibling of the path node; rather, it can be a
+/// position at a lower spot in the tree altogether.
+///
+pub struct PositionPath {
+    root: Position,
+    leaf: Position,
+    leaves_count: u64,
+}
+
+impl PositionPath {
+    pub fn new(root: Position, leaf: Position, leaves_count: u64) -> Self {
+        Self {
+            root,
+            leaf,
+            leaves_count,
+        }
+    }
+
+    pub fn iter(self) -> PositionPathIter {
+        PositionPathIter::new(self.root, self.leaf, self.leaves_count)
+    }
+}
+
+pub struct PositionPathIter {
+    rightmost_position: Position,
+    current_side_node: Option<Position>,
+    path_iter: PathIter<Position>,
+}
+
+impl PositionPathIter {
+    pub fn new(root: Position, leaf: Position, leaves_count: u64) -> Self {
+        Self {
+            rightmost_position: Position::from_leaf_index(leaves_count - 1),
+            current_side_node: None,
+            path_iter: root.as_path_iter(&leaf),
+        }
+    }
+}
+
+impl Iterator for PositionPathIter {
+    type Item = (Position, Position);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        while let Some((path, mut side)) = self.path_iter.next() {
+            // To determine if the position is in the tree, we observe that the highest in-order
+            // index belongs to the tree's rightmost leaf position (as defined by the `leaves_count`
+            // parameter) and that all path nodes will have an in-order index less than or equal to
+            // the in-order index of this rightmost leaf position.
+            // If a path position has an in-order index greater than that of the rightmost leaf
+            // position, it is invalid in the context of this tree and must be discarded. However,
+            // the corresponding side node is valid (or is the ancestor of a valid side node) and
+            // represents the side node of a deeper path position that is also valid (i.e., has an
+            // in-order index less than or equal to that of the rightmost leaf). We can save
+            // reference to it now so that we can generate the path node and side node pair later,
+            // once both nodes are encountered.
+            if path.in_order_index() <= self.rightmost_position.in_order_index() {
+                // If we previously encountered a side node corresponding to an invalid path node,
+                // we observe that the next valid path node always pairs with this side node. Once
+                // the path and side node have been paired, we continue to pair path and side nodes
+                // normally.
+                if self.current_side_node.is_some() {
+                    side = self.current_side_node.take().unwrap()
+                }
```

Comment:
nit: Could be replaced with `side = self.current_side_node.take().or(side);` so it's even more clear that we always leave  `self.current_side_node` empty.

Reply:
I like the idea of making the intent more clear. I ended up with a slightly modified version of what you suggested:
```rust
side = self.current_side_node.take().unwrap_or(side);
```
This is because `side` is a `Position`, not an `Option<Position>`, so I need to unwrap `current_side_node` at some point in order to do the assignment.